### PR TITLE
Swift 4.2 support

### PR DESCRIPTION
**Pull request by djbe**
This is needed to support Xcode 10 (beta).
The codebase should still be compatible with Swift 4.1 (Xcode 9).

Currently the `test_predefinedColor` tests fail, because the `iterateEnum(from:)` function in `IBEnum` is broken somehow in Swift 4.2.

Note: I didn't change the Travis file yet.

### CODE REVIEW

**Comment by JakeLin** [comment]
@djbe thanks for your PR. Since Xcode 10 is still in beta, we can merge it yet. We will do an update when Xcode 10 release public GM version.

**Comment by djbe** [comment]
@JakeLin This PR is compatible with Xcode 9, that's why I use `swift(>=4.1.50)` instead of `swift(>=4.2)`.

I've fully tested it with both Xcode versions locally, and it works (except for the `iterateEnum(from:)` mentioned above).

**Comment by JakeLin** [comment]
@djbe interesting, do you know why `iterateEnum(from:)` is broken? Only broken for the unit tests but it is OK for the Library and the Demo app? Or `iterateEnum(from:)` doesn't work in Swift 4.2 at all?

**Comment by tbaranes** [comment]
Didn't swift 4.2 add an equivalent to `iterateEnum`? It would be worth to check it out to see if we can use it directly.

**Comment by djbe** [comment]
@tbaranes They added `CaseIterable` yes, but that can only be added to the original declaration, and it's something that doesn't exist in Swift 4.1. If we were to use it, we'd have to duplicate whole enums to support it.

**Comment by djbe** [comment]
@JakeLin From a quick test, it's also broken in the example app: the `makeRandomGradient()` crashes  because the `predefinedGradients` collection is empty (because it can't iterate):
https://github.com/IBAnimatable/IBAnimatable/blob/master/IBAnimatableApp/IBAnimatableApp/Playground/Utils/UIViewControllerExtension.swift#L13

**Comment by djbe** [comment]
Huh, initial documentation seems to be wrong, I can add `CaseIterable` in extensions 🎉

**Comment by djbe** [comment]
Aha, so there is a limitation:
> Implementation of 'CaseIterable' cannot be automatically synthesized in an extension in a different file to the type

File: Sources/Common/SwiftSupport.swift
Comment (JakeLin) [inline Sources/Common/SwiftSupport.swift:10]: Great idea 👍

File: IBAnimatableApp/IBAnimatableApp/AppDelegate.swift
Comment (JakeLin) [inline IBAnimatableApp/IBAnimatableApp/AppDelegate.swift:43]: Shall we put those in one place?

File: Sources/Animators/TransitionAnimator/SystemTransitionAnimator.swift
Comment (JakeLin) [inline Sources/Animators/TransitionAnimator/SystemTransitionAnimator.swift:57]: Finally,  we have non-stringy type 🎉

**Review by JakeLin — COMMENTED** [review]
@djbe thanks for fixing that, the build is green now. I am still thinking shall we wait for a bit to merge it when Xcode 10 GM is out.

File: IBAnimatableApp/IBAnimatableApp/AppDelegate.swift
Comment (djbe) [reply IBAnimatableApp/IBAnimatableApp/AppDelegate.swift:43]: These are for the demo app, they shouldn't influence the code of the library itself.

**Comment by JakeLin** [comment]
@tbaranes @phimage how do you think? shall we wait for Xcode 10 GM and make a big release for it?

I am also thinking, shall we keep or remove those `#if swift(>=4.2)` code after Xcode 10 GM? Because most of the developer will upgrade their Xcode to version 10 once it is released publicly.

**Comment by djbe** [comment]
@JakeLin when to merge is up to you. It would just be useful for developers preparing their app for the next iOS/Xcode release to have this library be compatible with the new swift version. As stated before, the changes in this PR are **backwards compatible**.

I would not remove the `#if swift(>=4.2)` bits, as it would break the library for users with Xcode 9. Keep in mind that quite a large user base doesn't directly upgrade to Xcode 9, because of bugs, company policies, lack of time, etc...

**Comment by JakeLin** [comment]
@djbe I would like to consult our team member before I go ahead to merge it. We have a small Slack channel for our core developers, would you like to join us? If yes, please send me your Slack email to my email. You can find it on my profile page, thanks.

**Comment by tbaranes** [comment]
If it's compatible with swift 4.1, I'm completely ok to merge it right now! It will make the migration much easier.

**Review by JakeLin — APPROVED** [review]
👍 the first step to Swift 4.2 🎉

**Comment by JakeLin** [comment]
@djbe After discussed with team and they have done a few of tests, it looks good, let's go forward, thanks 👍

File: Sources/Common/SwiftSupport.swift
Comment (SD10) [reply Sources/Common/SwiftSupport.swift:10]: This could help me with another project. Thanks for the trick @djbe 💯

**Pull request merged**
